### DISCUSSION TITLE
Convergence plots

### DISCUSSION THREAD

**Discussion by shadowbourne** [wenceorg/comp4187#58] [Q&A] [answered]
I have a working implementation which when visually plotted converges to the correct graph.

However, when attempting to plot convergence of my solution as is shown in the theory and code from the lectures (the error (linalg.norm(x-fx, np.inf)) between my answer and the actual answer at final timesteps T), this error converges to a small value which is non-zero ( my time stepping scheme converges to a slightly different solution than the "real" solution, so the errors converge to 0.17 rather than 0).

This means that my convergence plots are flat  and my calculated convergence rate are close to 0.

However, if I calculate convergence in the method shown in psc1, then I obtain a convergence order of between 0.9 and 1 depending on grid resolution. (delta error for h**-i and h**-(i+1), rather than absolute error with analytical solution.)

Therefore I am confused whether I have done something wrong and we should be following the method set out in pc2 or we should be following the theory taught in psc one (which makes more sense to me as the "convergence order" should be the order of convergence to whatever solution it converges to, rather than whether or not the solution converges to the correct value?)

**Comment by wence-** [comment]
To check, you have some numerical solution, which you compute by discretising the operator on a fixed grid and then timestepping with a decreasing timestep (call this solution `x_h(t; dt)`). You expect that as you send `dt` to zero then the error `||x_h(t; dt) - x^*(t)||_inf` also goes to zero at an appropriate rate. To compute `x^*(t)` the true solution, I think that you must in some sense interpolate the continuous solution to the grid points (which I guess you do by evaluating it at the grid points).

I too think that this value should go to zero at a rate given by the choice of timestepping scheme.

I am not sure what the PSCI method of computing the convergence of a timestepping scheme was/is, can you expand on that methodology? Is it just running with a tiny timestep and using that as your "exact" solution?

What could be going wrong? One thing to check is to just plot the difference `x_h(t; dt) - x^*(t)` on the grid and see if you can see a pattern (e.g. does all the error pile up at a boundary? if so likely the boundary conditions are enforced incorrectly). You could also check what happens after just a single timestep: how do the computed and exact solutions diverge?

**Comment by shadowbourne** [comment]
errors after 1 timestep (2**-13 seconds): 
![Figure_1](https://user-images.githubusercontent.com/18665030/139909726-fd544bce-f127-4c7a-b372-253d348bb880.png)

error after 2: 
![Figure_1](https://user-images.githubusercontent.com/18665030/139910144-02fce2cd-2325-45df-af96-5f7d271362cc.png)

error after 2**12 timesteps: 
![Figure_1](https://user-images.githubusercontent.com/18665030/139909491-1b0d226b-7cc6-4f02-81c3-454cda917975.png)

Every timestep seems to add exactly a sin wave with 1e-8 magnitude to the error..........?

![Figure_1](https://user-images.githubusercontent.com/18665030/139907762-5db0015e-076a-4b3d-9548-ecf90c20179b.png)

Any ideas what could possibly be going wrong???

**Reply by wence-** [reply]
Have you run a convergence plot from "quite large" timesteps down to smaller ones? You might be running into roundoff issues with dt this small.

**Reply by shadowbourne** [reply]
I have, from the largest timesteps that my scheme is stable for
> (which is highly dependent upon the grid resolution chosen, which I dont understand to be honest but I'm assuming is supposed to be the case (higher N or smaller dx means smaller dt required for stability) 

to timesteps as small as 2**-20.

The log log graph is almost completely flat (if i fit Ch^p, p =0.02) for this plot as the errors converge with order 1 to a solution that is very slightly different than the actual u, so errors go from ~0.17 to ~0.16124 with order 1.

It make sense that this error is constant as the error graphs I show above show that the error introduced is proportional to the time that the scheme is run for rather than the time step size itself.

However, if I plot a graph of ( |u - u*| - ~0.16124 ), i get convergence order ~0.97 when fitting Ch^p. Which suggests round-off issues are not the issue as the graph is almost perfect order 1 convergence?

This is why if I follow the psc 1 method of plotting |u(T, h^-i) - u(T,h^-(i+1))| for i in range(8, 18) for example i get convergence order 1 as this is same as plotting ( |u - u*| - ~0.16124 )

**Reply by wence-** [reply]
OK, if you can briefly write this up in your discussion I think that will work. I had a look at your notebook and don't seen anything wrong here.

The one (potential) issue with plotting convergence of `|u(T, h^{-i}) - u(T, h^{-(i+1)})|` is that your numerical solution might be converging, but not to the correct solution! So you might have beautiful convergence, but to a nonsense result. (e.g. here was a very expensive example of this happening: https://www-users.cse.umn.edu/~arnold/disasters/sleipner.html)

**Comment by annereinarz** [comment, marked as answer]
If after 2^12 timesteps you've still only got an error of 0.01, then I would say everything is working well.

In this case the most likely culprit is your error norm as Lawrence has suggested. What are you using there?

**Reply by shadowbourne** [reply]
This is what I thought, but it means that my convergence graph shows 0 convergence order as this method does not converge to the correct solution but one that is very slightly different.

To calculate the error I calculate the solution at final timestep T u(T, dt) which is an array of length N (1/dx).

I then take error = np.linalg.norm( u(T, dt) - u*(T) , np.inf)

Calculate array of errors 1 entry for each dt.

If I plot this I get p = 0.02 and flat log log convergence.

However, if I plot errors[i+1] - errors[i] like we were shown in psc1 i get a proper order 1 convergence graph.
>  errors[i+1] - errors[i] In this case is equivalent to not knowing u* as the u*(T) cancels out

Does this answer your question? Or have I misunderstood.

**Reply by annereinarz** [reply]
That answers the question. I've also had a look at your notebook and I think the issue is the balancing of the two sources of error: you're introducing an error with your spatial discretisation and with your time discretisation. At some point the spatial error dominates and you don't "see" the convergence of your time discretisation anymore. By using the next finer approximation u(T, h^{-(i+1)}) you basically neglect that error term, so you can see the convergence again.

Try to change both time and space resolution together with h_t proportional to h_x^2/k. Then you should see the order 1 convergence even with comparison against the analytical solution.

**Reply by shadowbourne** [reply]
Thank you! Is there any theory I should know or quote as to why h_t should be decreased proportionally to h_x^2/k ?? Or is sufficient to say that this is simply what I did in my experiment?

**Reply by annereinarz** [reply]
You don't need it for the report, but it's essentially the theory we looked at in the lecture for the restriction on the time-step.

We had this restriction: |\lambda h + 1|\leq 1. Here everything is a bit more complicated than in the lecture since we're actually interested in a second derivative in time, but essentially this gives us that h_t^2 is bounded by the maximum eigenvalue of your spatial discretisation matrix, which behaves like k/h_x^4.

**Reply by shadowbourne** [reply]
Also to verify h_t should be changed proportional to h_x^(2/k) or (h_x^2)/k?

Thankyou!

**Reply by annereinarz** [reply]
(h_x^2)/k  and actually now that I've typed out the reasoning: I see it should be sqrt(k) instead of k. It's all up to constants though, so you have some leeway there.

Could you mark the question as resolved as well?

**Reply by shadowbourne** [reply]
I was literally just about to ask should it be sqrt(k) 😆  however the constant cancels out anyways i believe when doing the maths.
I'm about to run the code and then i shall mark the question resolved!
